### ISSUE TITLE
Rook/Ceph CSI 1.1.8 and later: Restore snapshot to PVC is pending forever

### ISSUE THREAD

**Issue by hseipp** [rook/rook#4624]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
PVC created from a ceph rbd snapshot in the `default` namespace forever remains in `Pending` state.

**Expected behavior:**
PVC that is created from a snapshot should be successfully `Bound` as [documented](https://rook.io/docs/rook/v1.2/ceph-csi-drivers.html). Note that this works with rook 1.1.7 on the same setup.

**How to reproduce it (minimal and precise):**
Checkout the rook release-1.1 Git branch or use rook version 1.1.8, then follow the steps in the [Documentation](https://rook.io/docs/rook/v1.2/ceph-csi-drivers.html). Note that the issue also occurs with rook 1.2.0.

**File(s) to submit**:

* Cluster CR (custom resource), typically called `cluster.yaml`, if necessary
Unmodified `cluster-test.yaml` is being used here.
* Operator's logs, if necessary
* Crashing pod(s) logs, if necessary
PVC status:
```
$ kubectl get pvc
NAME              STATUS    VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS      AGE
rbd-pvc           Bound     pvc-d6635ace-0949-43ca-9e02-072d602650a7   1Gi        RWO            rook-ceph-block   70m
rbd-pvc-restore   Pending                                                                        rook-ceph-block   70m
```

`rbd-pvc-restore` PVC log excerpt:
```  Type     Reason                Age                     From                                                                                                        Message
  ----     ------                ----                    ----                                                                                                        -------
  Normal   Provisioning          3m12s (x10 over 7m50s)  rook-ceph.rbd.csi.ceph.com_csi-rbdplugin-provisioner-54cc7d5848-rtsmd_4b7e3884-f040-4b29-a5c3-c2cb5bea1b9f  External provisioner is provisioning volume for claim "default/rbd-pvc-restore"
  Warning  ProvisioningFailed    3m10s (x10 over 7m47s)  rook-ceph.rbd.csi.ceph.com_csi-rbdplugin-provisioner-54cc7d5848-rtsmd_4b7e3884-f040-4b29-a5c3-c2cb5bea1b9f  failed to provision volume with StorageClass "rook-ceph-block": volume content source missing
  Normal   ExternalProvisioning  2m2s (x26 over 7m50s)   persistentvolume-controller                                                                                 waiting for a volume to be created, either by external provisioner "rook-ceph.rbd.csi.ceph.com" or manually created by system administrator
```
`csi-provisioner` Docker log excerpt:
```
I0107 14:32:54.942702       1 controller.go:824] VolumeContentSource_Snapshot {Snapshot:snapshot_id:"0001-0009-rook-ceph-0000000000000001-b196096c-3153-11ea-8753-36d32e3dab18" }
I0107 14:32:54.942732       1 controller.go:832] Requested volume size is 1073741824 and snapshot size is 1073741824 for the source snapshot rbd-pvc-snapshot
I0107 14:32:54.942741       1 controller.go:494] CreateVolumeRequest {Name:pvc-fe2d393b-c44e-42fc-910e-2943c68da771 CapacityRange:required_bytes:1073741824  VolumeCapabilities:[mount:<fs_type:"ext4" > access_mode:<mode:SINGLE_NODE_WRITER > ] Parameters:map[clusterID:rook-ceph csi.storage.k8s.io/fstype:ext4 csi.storage.k8s.io/node-stage-secret-name:rook-csi-rbd-node csi.storage.k8s.io/node-stage-secret-namespace:rook-ceph csi.storage.k8s.io/provisioner-secret-name:rook-csi-rbd-provisioner csi.storage.k8s.io/provisioner-secret-namespace:rook-ceph imageFeatures:layering imageFormat:2 pool:replicapool] Secrets:map[] VolumeContentSource:snapshot:<snapshot_id:"0001-0009-rook-ceph-0000000000000001-b196096c-3153-11ea-8753-36d32e3dab18" >  AccessibilityRequirements:<nil> XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}
I0107 14:32:54.944629       1 connection.go:180] GRPC call: /csi.v1.Controller/CreateVolume
I0107 14:32:54.944639       1 connection.go:181] GRPC request: {"capacity_range":{"required_bytes":1073741824},"name":"pvc-fe2d393b-c44e-42fc-910e-2943c68da771","parameters":{"clusterID":"rook-ceph","imageFeatures":"layering","imageFormat":"2","pool":"replicapool"},"secrets":"***stripped***","volume_capabilities":[{"AccessType":{"Mount":{"fs_type":"ext4"}},"access_mode":{"mode":1}}],"volume_content_source":{"Type":{"Snapshot":{"snapshot_id":"0001-0009-rook-ceph-0000000000000001-b196096c-3153-11ea-8753-36d32e3dab18"}}}}
I0107 14:32:56.238887       1 connection.go:183] GRPC response: {"volume":{"capacity_bytes":1073741824,"volume_context":{"clusterID":"rook-ceph","imageFeatures":"layering","imageFormat":"2","pool":"replicapool"},"volume_id":"0001-0009-rook-ceph-0000000000000001-97b407a3-315a-11ea-8753-36d32e3dab18"}}
I0107 14:32:56.239456       1 connection.go:184] GRPC error: <nil>
I0107 14:32:56.239469       1 controller.go:549] create volume rep: {CapacityBytes:1073741824 VolumeId:0001-0009-rook-ceph-0000000000000001-97b407a3-315a-11ea-8753-36d32e3dab18 VolumeContext:map[clusterID:rook-ceph imageFeatures:layering imageFormat:2 pool:replicapool] ContentSource:<nil> AccessibleTopology:[] XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}
I0107 14:32:56.239490       1 connection.go:180] GRPC call: /csi.v1.Controller/DeleteVolume
I0107 14:32:56.239494       1 connection.go:181] GRPC request: {"secrets":"***stripped***","volume_id":"0001-0009-rook-ceph-0000000000000001-97b407a3-315a-11ea-8753-36d32e3dab18"}
I0107 14:32:57.350657       1 connection.go:183] GRPC response: {}
I0107 14:32:57.351470       1 connection.go:184] GRPC error: <nil>
I0107 14:32:57.351541       1 controller.go:1023] Temporary error received, adding PVC fe2d393b-c44e-42fc-910e-2943c68da771 to claims in progress
W0107 14:32:57.351571       1 controller.go:887] Retrying syncing claim "fe2d393b-c44e-42fc-910e-2943c68da771", failure 18
E0107 14:32:57.351603       1 controller.go:910] error syncing claim "fe2d393b-c44e-42fc-910e-2943c68da771": failed to provision volume with StorageClass "rook-ceph-block": volume content source missing
I0107 14:32:57.351647       1 event.go:255] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"default", Name:"rbd-pvc-restore", UID:"fe2d393b-c44e-42fc-910e-2943c68da771", APIVersion:"v1", ResourceVersion:"15534", FieldPath:""}): type: 'Warning' reason: 'ProvisioningFailed' failed to provision volume with StorageClass "rook-ceph-block": volume content source missing
```

That following [commit](https://github.com/rook/rook/commit/319b2d211267cf39abbad089a93c22619863c360) (updated CSI sidecar images) might be related to the issue. The "volume content source missing" error message got introduced with the following [commit](https://github.com/kubernetes-csi/external-provisioner/pull/283) for the csi-provisioner sidecar.

**Environment**:
* OS (e.g. from /etc/os-release):
CentOS 7
* Kernel (e.g. `uname -a`):
3.10.0-957.12.2.el7.x86_64
* Cloud provider or hardware configuration:
1 master, 2 worker nodes, all KVM VMs
* Rook version (use `rook version` inside of a Rook Pod):
rook: v1.1.8
go: go1.11
* Storage backend version (e.g. for ceph do `ceph -v`):
ceph version 14.2.4 (75f4de193b3ea58512f204623e6c5a16e6c1e1ba) nautilus (stable)
* Kubernetes version (use `kubectl version`):
1.16.3
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift):
Plain k8s, installed using kubeadm
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/master/ceph-toolbox.html)):
HEALTH_OK

**Comment by scotty-g** [comment]
I am seeing the same issue while attempting to restore a snapshot. 

- Minikube v1.6.2 (MacOS 10.13.6)
- Kubernetes v1.17
- Rook: v1.2.1

Downgrading to Rook v1.1.7 gets everything working.

Thank you @hseipp for logging this issue, it allowed me to progress with my experimentation with rook/ceph.

**Comment by travisn** [comment]
@Madhu-1 This sounds related to a CSI change that was included in [v1.1.8](https://github.com/rook/rook/releases/tag/v1.1.8). Can you take a look?

**Comment by hseipp** [comment]
Thank you for looking into this issue. I think the csi-provisioner sidecar image [change](https://github.com/kubernetes-csi/external-provisioner/pull/283/commits/a3196715292dfc529d951e90f3d13c33806b332c) only surfaced the problem. Root cause seems to be that ceph-csi is not setting the [ContentSource](https://github.com/container-storage-interface/spec/blob/master/lib/go/csi/csi.pb.go#L1398) property in the [CreateVolume](https://github.com/ceph/ceph-csi/blob/master/pkg/rbd/controllerserver.go#L120) response.

**Comment by Madhu-1** [comment]
@hseipp Thanks for looking into it, as CSI spec says ContentSource is optional ceph-csi is not setting this value now, need spec update for the same please check https://github.com/container-storage-interface/spec/issues/404

**Comment by hseipp** [comment]
> @hseipp Thanks for looking into it, as CSI spec says ContentSource is optional ceph-csi is not setting this value now, need spec update for the same please check [container-storage-interface/spec#404](https://github.com/container-storage-interface/spec/issues/404)

@Madhu-1 Does this mean that the ceph-csi code stays as-is until the spec is changed?
Aren't we on the same page that
1. It is correct that the csi-provisioner checks the validity of the ContentSource field for a CreateVolume request that has VolumeSnapshot as DataSource.
2. The spec needs to get adjusted so that ContentSource is mandatory for CreateVolume requests with VolumeSnapshot DataSource

If so, why can't you go ahead and set the ContentSource field in ceph-csi for CreateVolume requests with VolumeSnapshot DataSource? This will even be compliant to the unchanged spec, pass the improved data validation and make Ceph CSI create PVC from Snapshot work again.

**Comment by Madhu-1** [comment]
> can't you go ahead and set the ContentSource field in ceph-csi for CreateVolume requests with VolumeSnapshot DataSource? This will even be compliant to the unchanged spec, pass the improved data validation and make Ceph CSI create PVC from Snapshot work again.

Fix already merged to master and will be part of ceph-csi v2.0.0 release

**Comment by hseipp** [comment]
Thank you @Madhu-1 that was quick! I should have checked the ceph-csi commit history. I'm looking forward to testing the ceph-csi v2.0.0 release.

**Comment by zops** [comment]
Have the same problem

rook：1.2.1
k8s:1.17
os：centos 7.7

log:
waiting for a volume to be created, either by external provisioner "rook-ceph.rbd.csi.ceph.com" or manually created by system administrator

**Comment by dotbalo** [comment]
I also encountered the same problem in the 1.2 version of rook-ceph, but I did not see how to deal with this problem? Are there any solved examples? I used the following example in `cluster/examples/kubernetes/ceph/csi/rbd` to create a snapshot. But the use of pvc-restore is still in the `pending `state.

````
[root@k8s-master01 rbd]# kubectl get pvc
NAME              STATUS    VOLUME   CAPACITY   ACCESS MODES   STORAGECLASS      AGE
rbd-pvc-restore   Pending                                      rook-ceph-block   25m

[root@k8s-master01 rbd]# kubectl describe pvc rbd-pvc-restore
  Normal   ExternalProvisioning  4m58s (x82 over 25m)  persistentvolume-controller                                                                                 waiting for a volume to be created, either by external provisioner "rook-ceph.rbd.csi.ceph.com" or manually created by system administrator
  Normal   Provisioning          3m59s (x13 over 25m)  rook-ceph.rbd.csi.ceph.com_csi-rbdplugin-provisioner-6b8b4d558c-xw4qn_f636c9e3-72a3-43a7-b927-82a08b471a0e  External provisioner is provisioning volume for claim "default/rbd-pvc-restore"
  Warning  ProvisioningFailed    3m47s (x13 over 24m)  rook-ceph.rbd.csi.ceph.com_csi-rbdplugin-provisioner-6b8b4d558c-xw4qn_f636c9e3-72a3-43a7-b927-82a08b471a0e  failed to provision volume with StorageClass "rook-ceph-block": volume content source missing
````

Versions:
````
[root@k8s-master01 rbd]# kubectl version
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.0", GitCommit:"70132b0f130acc0bed193d9ba59dd186f0e634cf", GitTreeState:"clean", BuildDate:"2019-12-07T21:20:10Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.0", GitCommit:"70132b0f130acc0bed193d9ba59dd186f0e634cf", GitTreeState:"clean", BuildDate:"2019-12-07T21:12:17Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"linux/amd64"}
[root@k8s-master01 rbd]# uname -a
Linux k8s-master01 4.18.0-80.el8.x86_64 #1 SMP Tue Jun 4 09:19:46 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
[root@k8s-master01 rbd]# cat /etc/redhat-release 
CentOS Linux release 8.0.1905 (Core)

[root@k8s-master01 rbd]# git status
On branch release-1.2
Your branch is up to date with 'origin/release-1.2'.
````

**Comment by hseipp** [comment]
@dotbalo With Rook 1.2 you need to manually switch to the ceph-csi 2.0.x driver along with its prereq csi-attacher and csi-provisioner pods. I once did this successfully to verify the above fix. I did not take notes, but if I recall correctly you need to uncomment the [following lines](https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/ceph/operator.yaml#L49) (all those with `ROOK_CSI_*_IMAGE`) before starting your deployment.
Also please be aware that with k8s 1.17 the [Volume Snapshot API changed](https://kubernetes.io/blog/2019/12/09/kubernetes-1-17-feature-cis-volume-snapshot-beta/), so csi-snapshotter 2.0.x would be required to comply with the new Beta CSI snapshot. ceph-csi seems not yet to be ready to consume this.
CSI snapshot Alpha might still be supported in 1.17 otherwise scotty-g would not have reported success above, but to be on the safe side, I'd rather stick with k8s 1.16.

**Comment by dotbalo** [comment]
@hseipp 
Thank you so much for your answers and suggestions. 
I changed the image to 2.0.0, and everything worked. But there is a small problem, I changed the `apiVersion `of `snapshotclass.yaml` to `apiVersion: snapshot.storage.k8s.io/v1beta1`, and there will be the following error when creating: `no matches for kind "VolumeSnapshotClass" in version "snapshot.storage.k8s.io/v1beta1"`. 
If it is v1alpha1 then everything is normal.

**Comment by Madhu-1** [comment]
Beta1 is not supported with current cephcsi version, we are planning to update snapshot to beta in cephcsi 3.0.0

**Comment by dotbalo** [comment]
@Madhu-1 Thank you very much. I will continue to pay attention.

**Comment by shaekhhasanshoron** [comment]
> before starting your deployment.

Hello, @hseipp I have uncommented all those with ROOK_CSI_*_IMAGE as you suggested. I am using rook v1.2.7 and k8 v1.17.1.
Currently, the PVC's doesn't stay in pending state after restoring from a snapshot. But now I am facing a different issue. Whenever I am creating a Statefulset, new PVC and Pv are generating successfully but pods stay in ContainerCreating state. I got the following issue while describing the pod. Can anyone please help me.

```
MountVolume.MountDevice failed for volume "pvc-dc3e1023-597e-444c-b3c1-e7e44bfcf01e" : rpc error: code = Aborted desc = an operation with the given Volume ID 0001-0009-rook-ceph-0000000000000001-21255c3a-b850-11ea-b60c-e218d8db998c already exists
```

**Comment by hseipp** [comment]
@shaekhhasanshoron As you mentioned, the issue you encounter is not related to this particular topic (restore from Snapshot failing). I'd suggest to create a new ticket and provide all information required to understand your issue (like the YAMLs for the statefulset and PVC you are referring to) there.
